common/util/hal_i2c_slave.c

```diff
@@ -0,0 +1,556 @@
+  Note: 
+    (1) Shall not modify code in this file!!!
+
+    (2) "hal_i2c_slave.h" must be included!
+
+    (3) User APIs follow check-rule before doing task 
+          [api]                               [.is_init] [.is_register]
+        * i2c_slave_control                   X          X
+        * i2c_slave_read                      O          X
+        * i2c_slave_status_get                X          X
+        * i2c_slave_status_print              X          X
+        * i2c_slave_cfg_get                   O          X
+                                              (O: must equal 1, X: no need to check)
+
+    (4) I2C slave function/api usage recommend
+        [ACTIVATE]
+          Use "i2c_slave_control()" to register/modify/unregister slave bus
+        [READ]
+          Use "i2c_slave_read()" to read slave queue message
+
+    (5) Slave queue method: Zephyr api, unregister the bus while full msgq, register back while msgq get space.
+*/
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <drivers/i2c.h>
+#include <sys/printk.h>
+#include "hal_i2c_slave.h"
+
+/* LOG SET */
+#include <logging/log.h>
+#define LOG_LEVEL CONFIG_I2C_LOG_LEVEL
+LOG_MODULE_REGISTER(mc_i2c_slave);
+
+/* I2C slave device arr */
+static struct i2c_slave_device i2c_slave_device_global[MAX_SLAVE_NUM] = { 0 };
+
+/* I2C slave config modify lock */
+struct k_mutex i2c_slave_mutex[MAX_SLAVE_NUM];
+
+/* static function declare */
+static int do_i2c_slave_cfg(uint8_t bus_num, struct _i2c_slave_config *cfg);
+static int do_i2c_slave_register(uint8_t bus_num);
+static int do_i2c_slave_unregister(uint8_t bus_num);
+
+static int i2c_slave_write_requested(struct i2c_slave_config *config)
+{
+  struct i2c_slave_data *data;
+
+  if (!config){
+    LOG_ERR("Get empty config!");
+    return 1;
+  }
+
+  data = CONTAINER_OF(config, struct i2c_slave_data, config);
+
+  data->current_msg.msg_length = 0;
+  memset(data->current_msg.msg, 0x0, MAX_I2C_SLAVE_BUFF);
+  data->buffer_idx = 0;
+
+  return 0;
+}
+
+static int i2c_slave_write_received(struct i2c_slave_config *config, uint8_t val)
+{
+  struct i2c_slave_data *data;
+  
+  if (!config){
+    LOG_ERR("Get empty config!");
+    return 1;
+  }
+
+  data = CONTAINER_OF(config, struct i2c_slave_data, config);
+
+  if (data->buffer_idx >= MAX_I2C_SLAVE_BUFF){
+    LOG_ERR("Buffer_idx over limit!");
+    return 1;
+  }
+  data->current_msg.msg[data->buffer_idx++] = val;
+
+  return 0;
+}
+
+static int i2c_slave_stop(struct i2c_slave_config *config)
+{
+  struct i2c_slave_data *data;
+  
+  if (!config){
+    LOG_ERR("Get empty config!");
+    return 1;
+  }
+
+  data = CONTAINER_OF(config, struct i2c_slave_data, config);
+
+  if (data->buffer_idx){
+    data->current_msg.msg_length = data->buffer_idx;
+
+#if DEBUG_MODE
+    printk("recv: ");
+    for (int i=0; i<data->current_msg.msg_length; i++){
+      printk("[%d]", data->current_msg.msg[i]);
+    }
+    printk("\n");
+#endif
```

Comment:
Is this debug code necessary or just left here from testing?
If not please remove it.

Reply:
Thanks for code review, it's for debug-use to check whether i2c slave ISR function received data correctly. We will remove it later update.

Reply:
Thanks for code review, it's for debug-use to check whether i2c slave ISR function received data correctly. We will remove it later update.